docs/notebooks/distributed/custom_loop_pmap_example.md

```diff
@@ -0,0 +1,335 @@
++++ {"id": "gbqilnNSNpw1"}
+
+# Auxiliary functions
+A MWE of how all-reduce mean/sum OPs can be introduced into JAXopt solver's `update` method by overriding `jax.value_and_grad`. Note that more complex wrappers are of course possible.
+
+```{code-cell}
+:id: awxxl498NwMI
+
+def pmean(fun: Callable[..., Any], axis_name: str = 'b') -> Callable[..., Any]:
+  """Applies `jax.lax.pmean` across `axis_name` for all of `fun`'s outputs."""
+  maybe_pmean = lambda t: jax.lax.pmean(t, axis_name) if t is not None else t
+  @functools.wraps(fun)
+  def wrapper(*args, **kwargs):
+    return jax.tree_map(maybe_pmean, fun(*args, **kwargs))
+  return wrapper
+```
+
++++ {"id": "clQeXb77ObhK"}
+
+A small utility to shard `Array`s across the available devices:
+
+```{code-cell}
+:id: rMntHIEFOeyO
+
+def shard_array(array: Array) -> jax.Array:
+  """Shards `array` along its leading dimension."""
+  return jax.device_put_sharded(
+      shards=list(common_utils.shard(array)),
+      devices=jax.devices())
+```
+
++++ {"id": "4Pph9LUAOmhi"}
+
+# Custom-loop
+Shows a comparison of how single-device and data-parallel SPMD differ in a vanilla JAXopt custom training loop
+
+```{code-cell}
+:id: tv1n_kkJO0Pf
+
+def fit(
+    data: Tuple[Array, Array],
+    init_params: Array,
+    stepsize: float = 0.0,
+    linesearch: str = 'zoom',
+    use_pmap: bool = False,
+) -> Tuple[np.ndarray, np.ndarray, float]:
+  """Fits a multi-class logistic regression model for demonstration purposes.
+
+  Args:
+    data: A tuple `(X, y)` with the training covariates and categorical labels,
+      respectively.
+    init_params: The initial value of the model's weights.
+    stepsize: The stepsize to use for the solver. If set to `0`, linesearch will
+      be used instead.
+    linesearch: The linesearch algorithm to use. If `stepsize > 0`, linsearch
+      will be disabled.
+    use_pmap: Whether to distribute the computation across replicas or use only
+      the first device available.
+  
+  Returns:
+    The per-step errors and runtimes, as well as the JIT-compile time for the
+    solver's `update` function.
+  """
+  # Value and grad of the objective function for the solver.
+  value_and_grad_fun = jax.value_and_grad(jaxopt.objective.multiclass_logreg)
+  # When `jax.pmap`ing the computation, use JAXopt's option to provide a custom
+  # `value_and_grad` function to include the desired reduction operators. For
+  # example, here we decide to average across replicas.
+  if use_pmap:
+    value_and_grad_fun = pmean(value_and_grad_fun)
+
+  # To override `jax.value_and_grad` in a JAXopt solver, set the flag 
+  # `value_and_grad` to `True` and pass the custom implementation of the
+  # `value_and_grad` function as `fun`.
+  solver = jaxopt.LBFGS(fun=value_and_grad_fun,
+                        value_and_grad=True,
+                        stepsize=stepsize,
+                        linesearch=linesearch)
+  # As usual, apply the `jax.pmap` transform to the function to be computed in
```

Comment:
I would remove "As usual", not sure to what it refers to (earlier in the script, general JAX idioms?)

Reply:
I was referring to it being arguably the most common usage pattern of `jax.pmap`, but fully agreed it doesn't bring anything here. Thanks!